### PR TITLE
DT-2408 Sentence Calc JpA

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java [hunk 1]
```diff
@@ -177,6 +193,167 @@
     @Column(name = "ADMISSION_REASON")
     private String admissionReason;
 
+    @OneToOne(mappedBy = "booking", cascade = CascadeType.ALL)
+    @PrimaryKeyJoinColumn
+    private ReleaseDetail releaseDetail;
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:198]: interesting this is a OneToOne, so for a booking with mutliple releases and recalls this just has the latest details ?

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java [hunk 2]
```diff
@@ -134,10 +138,22 @@
     @Default
     private List<SentenceCalculation> sentenceCalculations = new ArrayList<>();
 
+    @OneToMany(mappedBy = "offenderBooking", cascade = CascadeType.ALL)
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:141]: should these be lazy ?

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java [hunk 3]
```diff
@@ -177,6 +193,167 @@
...
 
+    @OneToOne(mappedBy = "booking", cascade = CascadeType.ALL)
+    @PrimaryKeyJoinColumn
+    private ReleaseDetail releaseDetail;
+
+    public Optional<SentenceCalculation> getLatestCalculation() {
+        return sentenceCalculations.stream().max(Comparator.comparing(SentenceCalculation::getId));
+    }
+
+    public LocalDate getReleaseDate() {
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:204]: Feels like thee should be unit for this

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java [hunk 4]
```diff
@@ -177,6 +193,167 @@
...
+     * offender, the release date is the later of the nonDtoReleaseDate or midTermDate value (if either or both
+     * are present).
+     *
+     * @param keyDateValues     a set of key date values used to determine the Non deterministic release date
+     * @param nonDtoReleaseDate derived Non deterministic release date information
+     * @return releaseDate
+     */
+    private static LocalDate deriveOffenderReleaseDate(final KeyDateValues keyDateValues, final NonDtoReleaseDate nonDtoReleaseDate) {
+
+        final LocalDate releaseDate;
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:314]: Feels like thee should be unit for this

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java [hunk 5]
```diff
@@ -177,6 +193,167 @@
...
+        return releaseDate;
+    }
+
+
+    public Optional<LocalDate> getConfirmedReleaseDate() {
+        return Optional.ofNullable(releaseDetail != null ? releaseDetail.getReleaseDate() != null ? releaseDetail.getReleaseDate() : releaseDetail.getAutoReleaseDate() : null);
+    }
+
+    public Optional<LocalDate> getSentenceStartDate() {
+        return sentences.stream()
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:345]: Feels like thee should be unit for this

File: src/main/java/uk/gov/justice/hmpps/prison/service/BookingService.java [hunk 6]
```diff
@@ -163,26 +168,69 @@ public BookingService(final BookingRepository bookingRepository,
...
+    public SentenceCalcDates getBookingSentenceCalcDatesV1_1(final Long bookingId) {
+
+        return offenderBookingRepository.findById(bookingId)
+            .orElseThrow(EntityNotFoundException.withId(bookingId))
+            .getSentenceCalcDates();
+    }
+
+    @NotNull
+    private SentenceCalcDates calcDerivedDates(final SentenceCalcDates sentenceDetail) {
+        final var derivedKeyDates = OffenderBooking.deriveKeyDates(buildKeyDateValues(sentenceDetail));
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/service/BookingService.java:202]: Feels like thee should be unit for this

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java (hunk 1 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:198]: Yes correct!

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java (hunk 2 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:141]: default is lazy for all OneToMany

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java (hunk 3 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:204]: There are lots of integration tests but no unit test other than NonDtoReleaseDateTest

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java (hunk 4 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java:314]: Release Dates is only tested as integration cucumber tests at the moment

**Pull request merged**